src/neurostatslib/glm.py

```diff
@@ -1,431 +1,598 @@
         """
-        try:
-            Ws = self.spike_basis_coeff_
-        except AttributeError:
-            raise NotFittedError(
-                "This GLM instance is not fitted yet. Call 'fit' with appropriate arguments."
-            )
-        bs = self.baseline_log_fr_
+        # check if the model is fit
+        self._check_is_fit()
+
+        # convert to jnp.ndarray
+        (coupling_basis_matrix,) = self._convert_to_jnp_ndarray(coupling_basis_matrix)
 
-        if init_spikes.shape[0] != bs.shape[0]:
+        n_basis_coupling = coupling_basis_matrix.shape[1]
+        n_neurons = self.baseline_link_fr_.shape[0]
+
+        if init_y is None:
+            init_y = jnp.zeros((coupling_basis_matrix.shape[0], n_neurons))
+
+        Wf = self.basis_coeff_[:, n_basis_coupling * n_neurons :]
+        Wr = self.basis_coeff_[:, : n_basis_coupling * n_neurons]
+        bs = self.baseline_link_fr_
+
+        feedforward_input, init_y = self.preprocess_simulate(
+            feedforward_input, params_f=(Wf, bs), init_y=init_y, params_r=(Wr, bs)
+        )
+
+        self._check_input_and_params_consistency(
+            (Wr, bs),
+            y=init_y,
+        )
+
+        if init_y.shape[0] != coupling_basis_matrix.shape[0]:
             raise ValueError(
-                "Number of neurons must be the same during simulation and fitting! "
-                f"init_spikes n_neurons: {init_spikes.shape[0]}, "
-                f"self.baseline_log_fr_ n_neurons: {self.baseline_log_fr_.shape[0]}"
+                "`init_y` and `coupling_basis_matrix`"
+                " should have the same window size! "
+                f"`init_y` window size: {init_y.shape[1]}, "
+                f"`coupling_basis_matrix` window size: {coupling_basis_matrix.shape[1]}"
             )
-        if init_spikes.shape[1] != self.spike_basis_matrix.shape[1]:
-            raise ValueError(
-                "init_spikes has the wrong number of time steps!"
-                f"init_spikes time steps: {init_spikes.shape[1]}, "
-                f"spike_basis_matrix window size: {self.spike_basis_matrix.shape[1]}"
+
+        subkeys = jax.random.split(random_key, num=feedforward_input.shape[0])
+        # (n_samples, n_neurons)
+        feed_forward_contrib = jnp.einsum("ik,tik->ti", Wf, feedforward_input)
+
+        def scan_fn(
+            data: Tuple[jnp.ndarray, int], key: jax.random.PRNGKeyArray
+        ) -> Tuple[Tuple[jnp.ndarray, int], Tuple[jnp.ndarray, jnp.ndarray]]:
+            """Scan over time steps and simulate activity and rates.
+
+            This function simulates the neural activity and firing rates for each time step
+            based on the previous activity, feedforward input, and model coefficients.
+            """
+            activity, chunk = data
+
+            # Convolve the neural activity with the coupling basis matrix
+            conv_act = convolve_1d_trials(coupling_basis_matrix, activity[None, :, :])[
```

Comment:
isn't `activity[None,:,:]` the same as `activity[None]`? using that hopefully gets this all on one line

Reply:
it's a bit counter-intuitive that this is the equivalent of adding an extra dimension at the beginning, should probably note that.

Reply:
yes, it's the same